tests/e2e/e2e_gov_test.go

```diff
@@ -0,0 +1,296 @@
+	s.writeGovCancelUpgradeSoftwareProposal(s.chainA)
+
+	s.T().Logf("Submitting Gov Proposal: Software Upgrade")
+	s.submitNewGovProposal(chainAAPIEndpoint, sender, proposalCounter, configFile("proposal_3.json"))
+	s.depositGovProposal(chainAAPIEndpoint, sender, fees.String(), proposalCounter)
+	s.voteGovProposal(chainAAPIEndpoint, sender, fees.String(), proposalCounter, "yes", false)
+
+	proposalCounter++
+
+	s.T().Logf("Submitting Gov Proposal: Cancel Software Upgrade")
+	s.submitNewGovProposal(chainAAPIEndpoint, sender, proposalCounter, configFile("proposal_4.json"))
+	s.depositGovProposal(chainAAPIEndpoint, sender, fees.String(), proposalCounter)
+	s.voteGovProposal(chainAAPIEndpoint, sender, fees.String(), proposalCounter, "yes", false)
+
+	s.verifyChainPassesUpgradeHeight(s.chainA, 0, proposalHeight)
+	s.T().Logf("Successfully canceled upgrade at height %d", proposalHeight)
+}
+
+/*
+fundCommunityPool tests the funding of the community pool on behalf of the distribution module.
+Test Benchmarks:
+1. Validation that balance of the distribution module account before funding
+2. Execution funding the community pool
+3. Verification that correct funds have been deposited to distribution module account
+*/
+func (s *IntegrationTestSuite) fundCommunityPool(chainAAPIEndpoint, sender string) {
+	s.Run("fund_community_pool", func() {
+		beforeDistUatomBalance, _ := getSpecificBalance(chainAAPIEndpoint, distModuleAddress, tokenAmount.Denom)
+		if beforeDistUatomBalance.IsNil() {
+			// Set balance to 0 if previous balance does not exist
+			beforeDistUatomBalance = sdk.NewInt64Coin(uatomDenom, 0)
+		}
+
+		s.execDistributionFundCommunityPool(s.chainA, 0, sender, tokenAmount.String(), fees.String())
+
+		// there are still tokens being added to the community pool through block production rewards but they should be less than 500 tokens
+		marginOfErrorForBlockReward := sdk.NewInt64Coin(uatomDenom, 500)
+
+		s.Require().Eventually(
+			func() bool {
+				afterDistPhotonBalance, err := getSpecificBalance(chainAAPIEndpoint, distModuleAddress, tokenAmount.Denom)
+				s.Require().NoErrorf(err, "Error getting balance: %s", afterDistPhotonBalance)
+
+				return afterDistPhotonBalance.Sub(beforeDistUatomBalance.Add(tokenAmount.Add(fees))).IsLT(marginOfErrorForBlockReward)
+			},
+			15*time.Second,
+			5*time.Second,
+		)
+	})
+}
+
+func (s *IntegrationTestSuite) verifyChainHaltedAtUpgradeHeight(c *chain, valIdx, upgradeHeight int) {
+	s.Require().Eventually(
+		func() bool {
+			currentHeight := s.getLatestBlockHeight(c, valIdx)
+
+			return currentHeight == upgradeHeight
+		},
+		30*time.Second,
+		5*time.Second,
+	)
+
+	counter := 0
+	s.Require().Eventually(
+		func() bool {
+			currentHeight := s.getLatestBlockHeight(c, valIdx)
+
+			if currentHeight > upgradeHeight {
+				return false
+			}
+			if currentHeight == upgradeHeight {
+				counter++
+			}
+			return counter >= 2
+		},
+		8*time.Second,
+		2*time.Second,
+	)
```

Comment:
If the chain started and the upgrade was successful before calling this method, the `currentHeight == upgradeHeight` will always be false. I think we can assume that if the chain no longer reaches the upgrade height in 30 seconds, the chain is halted. WDYT?

```suggestion
	s.Require().Eventually(
		func() bool {
			currentHeight := s.getLatestBlockHeight(c, valIdx)

			return upgradeHeight > currentHeight
		},
		30*time.Second,
		5*time.Second,
	)
```